### PR TITLE
Rework ZarrIO backend/Remove python3.7/Update HDMF and PyNWB min/Update workflows

### PR DESCRIPTION
**Pull request by mavaylon1**
## Motivation

What was the reasoning behind this change? Please explain the changes briefly.
This branch holds the re-work of the ZarrIO backend on how it handles object references. The changes are to follow the same workflow HDMF has when handling object references. 

With the changes to the backend, we have a temporary ignore on dtype conversion. It is has been raised in this ticket:  https://github.com/hdmf-dev/hdmf-zarr/pull/120

This PR removes python 3.7, brings hdmf and pynwb to the most up to date, as well updates to the workflows.

## How to test the behavior?
```
Show how to reproduce the new behavior (can be a bug fix or a new feature)
```

Run unit tests and gallery examples, specifically test_io_convert.py and plot_convert_nwb_hdf5.py 

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [x] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf-zarr/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR clearly describes the problem and the solution?
- [x] Is your contribution compliant with our coding style? This can be checked running `ruff` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf-zarr/pulls) for the same change?
- [x] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

File: src/hdmf_zarr/zarr_utils.py
Comment (oruebel) [inline src/hdmf_zarr/zarr_utils.py]:
```suggestion
Utilities for the Zarr I/O backend,
e.g., for wrapping Zarr arrays on read, wrapping arrays for configuring write, or
writing the spec among others
"""
```

File: src/hdmf_zarr/zarr_utils.py
Comment (oruebel) [inline src/hdmf_zarr/zarr_utils.py]: Add docstring for the class

File: src/hdmf_zarr/zarr_utils.py
Comment (oruebel) [inline src/hdmf_zarr/zarr_utils.py]: I think `BuilderResolverMixin` and `ContainerResolverMixin` are not implementing any Zarr-specific functionality. I think we may be able to just import these from HDMF. As the comment says, for import we may want to move them to `backend/utils.py` in HDMF instead

File: src/hdmf_zarr/zarr_utils.py
Comment (oruebel) [inline src/hdmf_zarr/zarr_utils.py:38]: Aside from inheriting from `ZarrDataset` I think the functionality in this class is the same as in `DatasetOfReferences` in  HDF5. We may want to think about how we can avoid this duplication.

File: src/hdmf_zarr/zarr_utils.py
Comment (oruebel) [inline src/hdmf_zarr/zarr_utils.py]: Add docstring for these classes.

For the classes for "Region" references, I would add a note that ZarrIO does not yet fully support region references and that the classes are, therefore, currently unused.

File: src/hdmf_zarr/backend.py
Comment (oruebel) [inline src/hdmf_zarr/backend.py]:
```suggestion
                if dts['dtype'] in ['object', 'region']: # check items for object reference
```

File: src/hdmf_zarr/backend.py
Comment (oruebel) [inline src/hdmf_zarr/backend.py]:
```suggestion
```

File: src/hdmf_zarr/backend.py
Comment (oruebel) [inline src/hdmf_zarr/backend.py]:
```suggestion
            if has_reference:
                # TODO:  BuilderZarrTableDataset does not yet support region reference
                data = BuilderZarrTableDataset(zarr_obj, self, retrieved_dtypes)
```

File: src/hdmf_zarr/backend.py
Comment (oruebel) [reply src/hdmf_zarr/backend.py]: For compound datasets, `BuilderZarrTableDataset` will need to handle the region references, so I think we can hear prepare the code to already wrap the dataset correctly so that if/when we add region reference supprort we will only need to update `BuilderZarrTableDataset`

File: src/hdmf_zarr/backend.py
Comment (oruebel) [inline src/hdmf_zarr/backend.py]:
```suggestion
            # TODO: Resolution of Region reference not yet supported by BuilderZarrRegionDataset
            elif dtype == 'region':
                data = BuilderZarrRegionDataset(data, self)
```

Comment (oruebel) [reply src/hdmf_zarr/backend.py]: You already have the wrapper classes for this setup, so I think we can do the wrapping here already. However, we should add a TODO item in the utils wherever region references need to be resolved and maybe raise a NotImplementedError if appropriate.

File: src/hdmf_zarr/backend.py
Comment (mavaylon1) [inline src/hdmf_zarr/backend.py]: @oruebel What Ryan and I found in HDMF was that the source is always the file itself and not an external source, in this case the builder.source is the file which it was written. The equivalent was what I did, which was to say `source = '.'`
Though HDF5 supports external references, we have references that are only internal.

This plus an update to resolve_ref was our fix until we add the "queue" mechanism of writing/exporting references in the future.

File: src/hdmf_zarr/backend.py
Comment (mavaylon1) [inline src/hdmf_zarr/backend.py]: @oruebel This is the update to `resolve_ref` I mentioned below. This checks that the relative_path works to retrieve the target. This would change/be updated as we bring zarr up to speed with the "queue" HDMF uses so that we don't need this check.

File: src/hdmf_zarr/zarr_utils.py
Comment (mavaylon1) [reply src/hdmf_zarr/zarr_utils.py:38]: Agreed. I think for now this is fine because I don't want to slow down a release on this next week on waiting for a HDMF ticket to be merged to support this change. I will make an issue ticket on HDMF still with the changes needed and a ticket on Zarr to make sure we update this once that has been merged. Timeline for that should be before Boston still as it shouldn't take long at all, that being said I don't want it slowing down this release. This is the same for the other comments you mentioned about refactoring to the backend in HDMF. @oruebel

File: src/hdmf_zarr/backend.py
Comment (oruebel) [reply src/hdmf_zarr/backend.py]: Can you clarify what you mean by "queue mechanism for export?

Comment (mavaylon1) [reply src/hdmf_zarr/backend.py]: @oruebel Currently in hdmf, references are resolved once everything else have been written. They are stored in a queue. 
```
@self.__queue_ref
                    def _filler():
                        self.logger.debug("Resolving object references and setting attribute on dataset '%s' "
                                          "containing attributes: %s"
                                          % (name, list(attributes.keys())))
                        refs = list()
                        for item in data:
                            refs.append(self.__get_ref(item))
                        dset = parent[name]
                        dset[()] = refs
                        self.set_attributes(dset, attributes)```
This guarantees that they will be resolved. This filler method will "fill in" on write with dset[()] = refs.

From my understanding, there is no set order to which objects are written, leaving the possibility that some references won't be properly resolved. Having them pushed to a queue to the end guarantees this resolution.

Comment (oruebel) [reply src/hdmf_zarr/backend.py]: Makes sense. Thanks for the clarification. I think this was probably added after ZarrIO was branched from HDF5IO, probably for export and that this has not made it in ZarrIO yet.

**Comment by mavaylon1** [comment]
Fix #121 
Fix #113
Fix #106 
Fix #79

**Comment by mavaylon1** [comment]
@rly I did the updates for using export_source for a flag; however, we would need to flush it out more with the export updates in the future. (specifically in links). The scope of these this PR is getting to large.

File: src/hdmf_zarr/backend.py
Comment (oruebel) [inline src/hdmf_zarr/backend.py]: For future reference, I would add a TODO item to clarify what needs to be done here in the future. I think all we really need is a way to figure out whether the target for the reference will be copied into the new file on export or whether it will remain only in the source we are exporting from. Otherwise, resolution of external links is I believe already being handled. 

```suggestion
        # TODO: This check assumes that all links are internal links on export. Need to deal with external links on export.
        if export_source is not None:
            source = '.'
```

File: src/hdmf_zarr/backend.py
Comment (oruebel) [inline src/hdmf_zarr/backend.py:362]: @mavaylon1 references are also stored in attributes. We will need to pass `export_source` also to `write_attributes` so that it can pass the `export_source` also in its calls to `__get_ref`

Comment (oruebel) [reply src/hdmf_zarr/backend.py:362]: You may want to search for `__get_ref` to make sure that all places where `__get_ref` is used on write are passed the `export_source` parameter.

File: CHANGELOG.md
Comment (oruebel) [inline CHANGELOG.md]:
```suggestion
### Enhancements
* Enhanced ZarrIO to resolve object references lazily on read similar to HDMF's `HDF5IO` backend @mavaylon1 [#120](https://github.com/hdmf-dev/hdmf-zarr/pull/120)

### Dependencies
* Updated HDMF and PyNWB version to the most recent release @mavaylon1 [#120](https://github.com/hdmf-dev/hdmf-zarr/pull/120)
* Updated minimum Python version from 3.7 to 3.8 @mavaylon1 [#120](https://github.com/hdmf-dev/hdmf-zarr/pull/120)
```

File: src/hdmf_zarr/backend.py
Comment (mavaylon1) [reply src/hdmf_zarr/backend.py:362]: Ah missed those. Good catch

File: src/hdmf_zarr/backend.py
Comment (oruebel) [inline src/hdmf_zarr/backend.py]: With the fix in `__get_ref` using `export_source`, this workaround on read should no longer be necessary

File: src/hdmf_zarr/backend.py
Comment (oruebel) [inline src/hdmf_zarr/backend.py]:
```suggestion
```

Comment is the same as the code below it

Comment (oruebel) [inline src/hdmf_zarr/backend.py]:
```suggestion
```

The first part of the `if` statement appears to refer to `object` references not `region` references

File: src/hdmf_zarr/backend.py
Comment (mavaylon1) [reply src/hdmf_zarr/backend.py]: @oruebel we do need this still. The source is correct, but there actual path includes the source. This is a fix to remove the source. This would be a bandage until the next update on the backend to flush out export, but also the paths for references.

Comment (oruebel) [reply src/hdmf_zarr/backend.py]: That seems like an issue that should be addressed in `__get_ref`. We are correcting the `source` in `__get_ref`. We'll probably need to do something similar for `path`. We shouldn't create files with bad paths.

**Comment by oruebel** [comment]
We should also update the note here https://github.com/hdmf-dev/hdmf-zarr/blob/dev/docs/source/storage.rst#region-references to mention that the new resolver classes will also need to be updated to implement region references

**Comment by mavaylon1** [comment]
> We should also update the note here https://github.com/hdmf-dev/hdmf-zarr/blob/dev/docs/source/storage.rst#region-references to mention that the new resolver classes will also need to be updated to implement region references

Done.

File: .github/workflows/run_all_tests.yml
Comment (rly) [inline .github/workflows/run_all_tests.yml]:
```suggestion
          conda install -c conda-forge tox
```
The importlib_metadata install from conda-forge was necessary under python 3.7 but is no longer needed if we drop testing on python 3.7

File: src/hdmf_zarr/backend.py
Comment (rly) [inline src/hdmf_zarr/backend.py]:
```suggestion
        # TODO: This check assumes that all links are internal links on export.
        # Need to deal with external links on export.
        if export_source is not None:
            # Make sure the source of the reference is now towards the new file
            # and not the original source when exporting.
            source = '.'
        else:
            source = os.path.relpath(os.path.abspath(source), start=self.abspath)
```

File: tests/unit/test_io_convert.py
Comment (rly) [inline tests/unit/test_io_convert.py]: Can we evaluate the classes are equal instead of the class name? That's more typical. This formulation makes it look like there is a reason we cannot import these classes. e.g.,
`self.assertIsInstance(exported_container.baz_data.data, ContainerH5ReferenceDataset)`

**Comment by rly** [comment]
This looks good to me. I'll let @oruebel do the final approval.

**Comment by rly** [comment]
Thanks for your hard work on this!

File: .readthedocs.yaml
Comment (oruebel) [inline .readthedocs.yaml:29]: Just curious. Why is specifying `path` necessary?

Comment (mavaylon1) [reply .readthedocs.yaml:29]: What ended up happening was the build was failing. I looked at hdmf to see how it was set up there. Looking at the error, it was not properly installing the requirements when building the docs. I looked at the documentation for how to setup this yaml and it says the "path" would be the path to the requirements file. I confirmed that in hdmf it was also present. These changes allowed for the docs to properly find the requirements and be built.

File: CHANGELOG.md
Comment (oruebel) [inline CHANGELOG.md]:
```suggestion
## 0.4.0 (Upcoming)
```

This seems more than just a bug-fix release. So I would recommend bumping the version to 0.4

File: docs/gallery/plot_convert_nwb_hdf5.py
Comment (oruebel) [inline docs/gallery/plot_convert_nwb_hdf5.py:108]: @mavaylon1 can you please create an issue for this so we don't forget to check on the dtype conversion issue (and include links to the lines where the issue is being suppressed)

File: docs/source/storage.rst
Comment (oruebel) [inline docs/source/storage.rst]: @mavaylon1  can you please move this line to the list of TODO items in the warning message right below it and add mention where the reference resolver classes are located.

File: src/hdmf_zarr/backend.py
Comment (oruebel) [inline src/hdmf_zarr/backend.py:589]: Just FYI, looking through issues, we already have an open issue on export with external links here https://github.com/hdmf-dev/hdmf-zarr/issues/49  I.e., any specific issues we identify here related to export with external links, we should document there. I added a note, about this particular change in `__get_ref` to the issue, but if there are other items that you identified in this PR, then please add notes to that issue.

File: src/hdmf_zarr/backend.py
Comment (oruebel) [inline src/hdmf_zarr/backend.py:832]:
```suggestion
            # TODO Region references are not yet support, but here how the code should look
            #  if isinstance(data, RegionBuilder):
            #      shape = (1,)
            #      type_str = 'region'
            #      refs = self.__get_ref(data.builder, data.region)
            if isinstance(data, ReferenceBuilder):
```

I know I originally suggested removing this comment. But I would keep it, because it will make implementing region references easier in the future.

File: src/hdmf_zarr/backend.py
Comment (oruebel) [inline src/hdmf_zarr/backend.py]: Very minor. I would move the function to make sure `get_container` and `get_builder` are in the same place in the code, since the two are related.

File: src/hdmf_zarr/zarr_utils.py
Comment (oruebel) [reply src/hdmf_zarr/zarr_utils.py:38]: Agreed, creating issues for this is fine.

File: tests/unit/utils.py
Comment (oruebel) [inline tests/unit/utils.py]: For future proofing, I would just use the ROOT_NAME from ZarrIO (which I believe is already being imported). Similarly, in the docval, the default for 'name' can then be ROOT_NAME as imported from ZarrIO instead of redefining 'root' as a string.

Comment (mavaylon1) [reply tests/unit/utils.py]: I don't see that we import ZarrIO to this file. This is also set in FooFile (like in HDMF), I will define it outside the class and use ROOT_NAME

File: docs/gallery/plot_convert_nwb_hdf5.py
Comment (mavaylon1) [reply docs/gallery/plot_convert_nwb_hdf5.py:108]: Done

File: tests/unit/test_io_convert.py
Comment (oruebel) [inline tests/unit/test_io_convert.py:187]: @mavaylon1 according to Codecov, this code seems to not be hit during the tests. It seems strange that the references appear as not tested with "ZARR" as a the target format. 

Screen Shot 2023-09-29 at 3 43 10 PM

Can you please double-check why this is not being used? The CodeCov for this PR is here https://app.codecov.io/gh/hdmf-dev/hdmf-zarr/pull/120

File: tests/unit/test_io_convert.py
Comment (oruebel) [inline tests/unit/test_io_convert.py]:
```suggestion
    TARGET_FORMAT = "ZARR"
```

@mavaylon1  based on the code above, I believe this should be "ZARR" (all uppercase). I think this is likely why the check for references does not appear to be run in the tests.

File: tests/unit/utils.py
Comment (oruebel) [inline tests/unit/utils.py]: Is there a reason why we would not just import it?

```suggestion
from hdmf_zarr.backend import ROOT_NAME
```

File: tests/unit/test_io_convert.py
Comment (mavaylon1) [reply tests/unit/test_io_convert.py]: Done

File: tests/unit/utils.py
Comment (mavaylon1) [reply tests/unit/utils.py]: I don't think so. I think the manual definition within utils.py is a legacy carry over from HDMF, where it is set with all the Foo and Baz containers and not imported.

File: tests/unit/test_io_convert.py
Comment (oruebel) [inline tests/unit/test_io_convert.py]: Same as above. I believe this needs to be "ZARR" (all caps) for the checks for references to be run.

```suggestion
    TARGET_FORMAT = "ZARR"
```

**Review by oruebel — APPROVED** [review]
Looks good to me. Thanks for your patients and hard work on this PR.

**Pull request merged**
